### PR TITLE
Fix creation of OVA server pod under enforced restricted namespaces

### PR DESCRIPTION
**Pull request by bkhizgiy**
When a new namespace is created in OpenShift, it is automatically labeled with `pod-security.kubernetes.io/audit`, which refers to a secondary, restricted security level. The situation diffrent in Kubernetes and for the default namespace, where such automatic labeling does not occur. When a namespace is marked as restricted, an additional label, `pod-security.kubernetes.io/enforce`, is added. This label represents the highest security level, where violations are not permitted. When deploying an OVA server pod, it is created with standard settings across all namespaces, except those marked only with the enforce option. In these cases, extra configuration is required to eliminate any potential security violations.

### CODE REVIEW

File: pkg/controller/provider/ova-setup.go [hunk 1]
```diff
@@ -185,47 +189,83 @@ func (r *Reconciler) createServerService(provider *api.Provider, ctx context.Con
...
 	}
+
+	podSpec := core.PodSpec{
+		Containers: []core.Container{container},
+		Volumes:    []core.Volume{volume},
+	}
+	return podSpec
+}
+
+func (r *Reconciler) checkAndUpdateNamespaces(namespaceName string) bool {
```
Comment (bennyz) [inline pkg/controller/provider/ova-setup.go:251]: suggestion:
```go
func (r *Reconciler) checkAndUpdateNamespaces(namespaceName string) bool {
	ns := core.Namespace{}
	err := r.Get(context.TODO(), client.ObjectKey{Name: namespaceName}, &ns)
	if err != nil {
		r.Log.Error(err, "Error getting namespace for restriction check")
		return false
	}

	commonLabel, commonExists := ns.Labels[enforceRestrictedLable]
	ocpLabel, ocpExists := ns.Labels[auditRestrictedLable]

	return commonExists && commonLabel == "restricted" && !(ocpExists && ocpLabel == "restricted")
}
```

File: pkg/controller/provider/ova-setup.go [hunk 2]
```diff
@@ -12,14 +12,18 @@ import (
...
-	ovaImageVar         = "OVA_PROVIDER_SERVER_IMAGE"
-	nfsVolumeNamePrefix = "nfs-volume"
-	mountPath           = "/ova"
-	pvSize              = "1Gi"
+	ovaServer              = "ova-server"
+	ovaImageVar            = "OVA_PROVIDER_SERVER_IMAGE"
+	nfsVolumeNamePrefix    = "nfs-volume"
+	mountPath              = "/ova"
+	pvSize                 = "1Gi"
+	auditRestrictedLable   = "pod-security.kubernetes.io/audit"
```
Comment (liranr23) [inline pkg/controller/provider/ova-setup.go:24]:
```suggestion
	auditRestrictedLabel   = "pod-security.kubernetes.io/audit"
```

File: pkg/controller/provider/ova-setup.go [hunk 3]
```diff
@@ -12,14 +12,18 @@ import (
...
+	enforceRestrictedLable = "pod-security.kubernetes.io/enforce"
```
Comment (liranr23) [inline pkg/controller/provider/ova-setup.go:25]:
```suggestion
	enforceRestrictedLabel = "pod-security.kubernetes.io/enforce"
```

File: pkg/controller/provider/ova-setup.go [hunk 4]
```diff
@@ -185,47 +189,83 @@ func (r *Reconciler) createServerService(provider *api.Provider, ctx context.Con
...
+
+	restricted := r.checkAndUpdateNamespaces(providerNamespace)
+	if restricted {
+		seccompProfile := &core.SeccompProfile{
+			Type: core.SeccompProfileTypeRuntimeDefault,
+		}
+		container.SecurityContext.RunAsUser = &user
+		container.SecurityContext.RunAsNonRoot = &nonRoot
+		container.SecurityContext.SeccompProfile = seccompProfile
+	}
```
Comment (liranr23) [inline pkg/controller/provider/ova-setup.go:233]: i would do it before the container creation, setting it straight into the `commonSecurityContext`. Or, dropping `commonSecurityContext`, setting it in the container and then doing this part as it is.

File: pkg/controller/provider/ova-setup.go (hunk 1 above)
Comment (liranr23) [inline pkg/controller/provider/ova-setup.go:251]: what does it updates? namespaces?
how about `isEnforcedRestrictionNamespame`?

File: pkg/controller/provider/ova-setup.go [hunk 5]
```diff
@@ -185,47 +189,83 @@ func (r *Reconciler) createServerService(provider *api.Provider, ctx context.Con
...
+		Name: namespaceName,
+	}
+	err := r.Get(context.TODO(), key, &ns)
+	if err != nil {
+		r.Log.Error(err, "Error getting namespace for restriction check")
+		return false
+	}
+
+	//common label for restricted namesapce
+	if commonLabel, exists := ns.Labels[enforceRestrictedLable]; exists && commonLabel == "restricted" {
```
Comment (liranr23) [inline pkg/controller/provider/ova-setup.go:263]: `restricted` is here twice, you can pull it out as a const.

**Review by liranr23 — COMMENTED** [review]
Did you check other pods we create(virt-v2v, populators)? In my opinion they work in the default namespace. I don't remember we ever checked the restriction level. Can the OVA provider work with the additional restriction here without checking the namespace?

File: pkg/controller/provider/ova-setup.go (hunk 1 above)
Comment (bkhizgiy) [reply pkg/controller/provider/ova-setup.go:251]: yes my mistake I changed the name and forgot to push the change

**Comment by bkhizgiy** [comment]
After testing it with virt-v2v pod its working without the extra configuration, but it seams that the difference relays on the fact the ova server runs as a deployment while the virt-v2v created directly as a pod. for the deployment in the highest security level  there is extra enforcement for the user that can run the pod while for the others its not required and determine automatically.

File: pkg/controller/provider/ova-setup.go [hunk 6]
```diff
@@ -185,47 +189,75 @@ func (r *Reconciler) createServerService(provider *api.Provider, ctx context.Con
...
 	}
+
+	podSpec := core.PodSpec{
+		Containers: []core.Container{container},
+		Volumes:    []core.Volume{volume},
+	}
+	return podSpec
+}
+
+func (r *Reconciler) isEnforcedRestrictionNamespame(namespaceName string) bool {
```
Comment (liranr23) [inline pkg/controller/provider/ova-setup.go:251]:
```suggestion
func (r *Reconciler) isEnforcedRestrictionNamespace(namespaceName string) bool {
```
i had a typo in my comment

Comment (bkhizgiy) [reply pkg/controller/provider/ova-setup.go]: haha just noticed, uploaded new version:)

**Comment by liranr23** [comment]
> After testing it with virt-v2v pod its working without the extra configuration, but it seams that the difference relays on the fact the ova server runs as a deployment while the virt-v2v created directly as a pod. for the deployment in the highest security level there is extra enforcement for the user that can run the pod while for the others its not required and determine automatically.

so, can we do it as restricted all the time on any namespace?

**Comment by bkhizgiy** [comment]
> > After testing it with virt-v2v pod its working without the extra configuration, but it seams that the difference relays on the fact the ova server runs as a deployment while the virt-v2v created directly as a pod. for the deployment in the highest security level there is extra enforcement for the user that can run the pod while for the others its not required and determine automatically.
> 
> so, can we do it as restricted all the time on any namespace?

No, for the audit security level when specifying the user + not root the pod creation being blocked,  plus I've tried to specify what is the the permitted UID range as part of the controller SCC  but its allowed for use only when setting the security to `enforce` (so same behavior).

**Review by liranr23 — APPROVED** [review]

**Pull request merged**